compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -3849,22 +3849,9 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
         end implicitArgs
 
         val args = implicitArgs(wtp.paramInfos, 0, pt)
-
-        def propagatedFailure(args: List[Tree]): Type = args match {
-          case arg :: args1 =>
-            arg.tpe match {
-              case ambi: AmbiguousImplicits =>
-                propagatedFailure(args1) match {
-                  case NoType | (_: AmbiguousImplicits) => ambi
-                  case failed => failed
-                }
-              case failed: SearchFailureType => failed
-              case _ => propagatedFailure(args1)
-            }
-          case Nil => NoType
-        }
-
-        val propFail = propagatedFailure(args)
+        val firstAmbiguous = args.tpes.find(_.isInstanceOf[AmbiguousImplicits])
+        def firstError = args.tpes.find(_.isError)
```

Comment:
That's worst case $\mathcal{O}(2n)$  in the length of `args` instead of $\mathcal{O}(n)$. I think the readability gain is worth it.

Reply:
Note: as before, I am only checking for `AmbiguousImplicits`. Should `TooUnspecific` fall in the same case? I am asking because both are considered `ambiguous` here:

https://github.com/lampepfl/dotty/blob/4c1d58f067f40c9cd52628f195c6306b5a483d4f/compiler/src/dotty/tools/dotc/typer/Implicits.scala#L442